Add initial tests for removeTrend & NoisyChannels

a-hurst opened the pull request:
<!--Thanks for contributing-->
<!--If this is your first time, please make sure to read the contributing guideline-->
<!--https://github.com/sappelhoff/pyprep/blob/master/.github/CONTRIBUTING.md-->

# PR Description

Closes #49. This PR aims to add direct CI comparison of PyPREP's `matlab_strict`results to Matlab PREP's across every stage of the pipeline, verifying that we're not deviating from MatPREP in any ways we're not actively aware of.

This is currently in a draft state, with only removeTrend and NoisyChannels being tested for equivalence, but I figure this would be a useful stage at which to get some feedback.

### Components currently tested

- [x] removeTrend
- [ ] CleanLine (`mne.filter.notch_filter`)
- [x] NoisyChannels
   - [x] Bad by NaN
   - [x] Bad by flat
   - [x] Bad by deviation
   - [x] Bad by HF noise
   - [x] Bad by correlation
   - [x] Bad by low SNR
   - [x] Bad by dropout
   - [x] Bad by RANSAC
- [ ] Robust re-referencing
- [ ] Final bad channel interpolation
- [ ] Full pipeline, start-to-finish

### Current thoughts

- The current test file I chose apparently only has two bad channels (T10 and FT8), none of which are bad by NaN, flat, SNR, or deviation. One possibility would be to try and find another test file that's noisier, another would be to try and simulate these types of noise as part of our `matlab_artifacts` script. Simulating bad-by-SNR is likely more trouble than just finding a noisier file, but I haven't found any data with bad-by-NaN or bad-by-dropout channels yet so I think simulation is the way to go for those.
- What do we want to do about the `examples/run_full_prep.py` script? Once #58 is finally finished it'll need to be rewritten a bit, but should we switch to using `matprep_artifacts` as the test data for that as well?
- How do we want to approach merging this? Should I write tests for the whole thing and mark the ones that aren't numerically-equivalent yet as expected failures, or should we merge once NoisyChannels tests are complete and add to this with future PRs when adding `matlab_strict` modes for things like re-referencing and final interpolation?

# Merge Checklist

<!--To merge your PR we need to first take the following points into account.-->
<!--Please just leave this checklist untouched-->

- [x] the PR has been reviewed and all comments are resolved
- [x] all [CI][what-is-ci] checks pass
- [x] (if applicable): the PR description includes the phrase `closes #<issue-number>` to [automatically close an issue][auto-close-documentation]
- [ ] (if applicable): bug fixes, new features, or [API][what-is-api] changes are documented in [whats_new.rst][whats-new-file]


[what-is-ci]: https://help.github.com/en/actions/building-and-testing-code-with-continuous-integration/about-continuous-integration
[auto-close-documentation]: https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
[what-is-api]: https://en.wikipedia.org/wiki/Application_programming_interface
[whats-new-file]: https://github.com/sappelhoff/pyprep/blob/master/docs/whats_new.rst

sappelhoff reviewed (commented):
Thanks for the start :)

> . One possibility would be to try and find another test file that's noisier, another would be to try and simulate these types of noise as part of our matlab_artifacts script.

I think testing 2 to 3 files instead of just one wouldn't hurt -> do you think that's feasible? 

> but I haven't found any data with bad-by-NaN or bad-by-dropout channels

I am not aware of such data either: just editing on of our datasets to contain NaN / dropout (whatever that is supposed to mean) might be fine enough?

> What do we want to do about the examples/run_full_prep.py script? Once #58 is finally finished it'll need to be rewritten a bit, but should we switch to using matprep_artifacts as the test data for that as well?

Yes, we should switch to matprep data there as well - and the plotting code needs to be a little bit less "in your face"

> How do we want to approach merging this? Should I write tests for the whole thing and mark the ones that aren't numerically-equivalent yet as expected failures, or should we merge once NoisyChannels tests are complete and add to this with future PRs when adding matlab_strict modes for things like re-referencing and final interpolation?

I think doing separate PRs would be a better way to go here. Add now what we can assure passes, and do the rest later

a-hurst commented:
@sappelhoff I've now switched `matprep_artifacts` to using a noisier test file, and added code in my MATLAB script for simulating NaN, flat, and dropout channels. NoisyChannels show now have full MATLAB-comparison test coverage!

Due to differences in how PyPREP and MatPREP generate their arrays of internal info, I also had to refactor `NoisyChannels` a bit to get this working properly with bad-by-NaN/bad-by-flat channels. Basically, PyPREP was dropping the bad-by-flat/bad-by-NaN channels completely such that for a dataset with two NaN channels, the arrays channel noisiness z-scores, RANSAC correlations, etc. would be 62 channels long/wide instead of 64 (which is how MatPREP outputs these arrays). Relatedly, because NaN and flat channels break the other `bad_by` detection methods in various spectacular ways, I got `NoisyChannels` to automatically run it during init.

Also I'm not sure what to do about the style check throwing the error: for that line, `is False` does something completely different than `== False` since it's comparing multiple elements in an array.

sappelhoff commented on pyprep/find_noisy_channels.py:
```diff
@@ -90,6 +79,25 @@ def __init__(self, raw, do_detrend=True, random_state=None, matlab_strict=False)
+        bads_by_nan_flat = self.bad_by_nan + self.bad_by_flat
+
+        # Make a subset of the data containing only usable EEG channels
+        self.usable_idx = np.isin(ch_names, bads_by_nan_flat) == False
```
```suggestion
        self.usable_idx = np.isin(ch_names, bads_by_nan_flat).astype(int) == 0
```

a-hurst replied:
Apparently `np.isin` has an `inverse` option that does what we want, I just missed it the first time around. Should be fixed now!

sappelhoff replied:
much nicer! thanks @a-hurst

a-hurst commented:
I think codecov's upset I uncommented an untested line in `find_bad_by_hf_noise`. I added a quick test case to make it happy.

sappelhoff reviewed (approved):
awesome work, thanks heaps :)

I think this is worth a big What's New entry, WDYT? -> or do you want to hold off with that until all things PREP are tested in CI?

Regarding the latter, where are we going to track the as of yet not tested in CI parts of PYPREP (to be done in later PRs)?

a-hurst commented:
Just updated the `whats_new.rst` to say we've added a framework for MATLAB comparison :)

For tracking the rest of the comparisons, I'd say we can use the checklist in #58 as a rough test coverage tracker, since it's essentially in the same state and I can add MATLAB-comparison tests as the remaining boxes in there get checked off.

sappelhoff commented:
Great, thanks!
